app/main/lib/elastic_crud.py

```diff
@@ -0,0 +1,99 @@
+import copy
+import uuid
+import json
+from flask import current_app as app
+from app.main.lib.presto import Presto, PRESTO_MODEL_MAP
+from app.main.lib.elasticsearch import store_document, get_by_doc_id
+
+def _after_log(retry_state):
+    app.logger.debug("Retrying image similarity...")
+
+def get_object_by_doc_id(doc_id):
+    return get_by_doc_id(doc_id)
+
+def get_object(task, _):
+    doc_id = task.get("doc_id", None)
+    language = task.get("language", None)
+    context = task.get("context", {})
+    if context:
+        task["contexts"] = [context]
+    store_document(task, doc_id, language)
+    if task.get("content") and not task.get("text"):
+        task["text"] = task["content"]
+    return task, False
+
+def get_context_for_search(task):
+    context = {}
+    dup = copy.deepcopy(task)
+    if dup.get('context'):
+        context = dup.get('context')
+    if dup.get("match_across_content_types"):
+        context.pop("content_type", None)
+    return context
+
+def get_presto_request_response(modality, callback_url, task):
+    response = json.loads(Presto.send_request(app.config['PRESTO_HOST'], PRESTO_MODEL_MAP[modality], callback_url, task, False).text)
+    assert response["message"] == "Message pushed successfully", f"Bad response message for {modality}, {callback_url}, {task} - response was {response}"
+    assert response["queue"] in PRESTO_MODEL_MAP.values(), f"Unknown queue for {modality}, {callback_url}, {task} - response was {response}"
+    assert isinstance(response["body"], dict), f"Bad body for {modality}, {callback_url}, {task} - response was {response}"
+    return response
+
+def requires_encoding(obj):
+    for model_key in obj.get("models", []):
+        if not obj.get('model_'+model_key):
+            return True
+    return False
+
+def get_blocked_presto_response(task, model, modality):
+    if task.get("doc_id") is None:
+        task["doc_id"] = str(uuid.uuid4())
+    obj, temporary = get_object(task, model)
+    doc_id = obj["doc_id"]
+    callback_url = Presto.add_item_callback_url(app.config['ALEGRE_HOST'], modality)
+    app.logger.info(f"Object for {task} of model {model} with id of {doc_id} has requires_encoding value of {requires_encoding(obj)}")
+    if requires_encoding(obj):
+        blocked_results = []
+        for model_key in obj.pop("models", []):
+            if model_key != "elasticsearch" and not obj.get('model_'+model_key):
+                response = get_presto_request_response(model_key, callback_url, obj)
```

Comment:
We need to handle OpenAI embeddings at some place. Will these have a worker in Pesto or will Alegre call OpenAI directly?

Reply:
Good point. I see good arguments for both ways here - if you send that work to presto, you're centralizing concerns, not writing more code here, but adding some time to the work. If you do the opposite, vice versa. What do you think @computermacgyver?

Reply:
I'd probably call OpenAI in Alegre. While it's a bit more code here, it is far less code than it would be on Presto's side to have a separate worker, queues, etc.